scripts/dts: Fix detection of a GPIO specifier

galak opened the pull request:
We should only assume a GPIO specifier is either named <FOO>-gpios or
gpios.  Any other form like ngpios should not be considered a GPIO
specifier.  especially since 'ngpios' is the standard property name for
the number of gpio's that are implemented.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts/dts: Fix detection of a GPIO specifier

We should only assume a GPIO specifier is either named <FOO>-gpios or
gpios.  Any other form like ngpios should not be considered a GPIO
specifier.  especially since 'ngpios' is the standard property name for
the number of gpio's that are implemented.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -154,8 +154,8 @@ def write_props(dev):
-        if prop.name in {"reg", "interrupts", "pwms", "clocks", "compatible"} or \
-           prop.name.endswith("gpios"):
+        if prop.name in {"reg", "interrupts", "pwms", "clocks", "compatible", "gpios"} or \
+           prop.name.endswith("-gpios"):
```
Looking at it now, I realized this can just be

```python
if prop.name in {"reg", "interrupts", "compatible"}:
    continue
```

, because we never create `Property` objects for properties declared with `type: compound`.

If someone declares e.g. `foo-gpios` as an `int`, then something else is probably broken with their PR, and the check won't do any good in those cases either.

Could you change that while we're here?

galak replied:
Done.

ulfalizer commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -84,7 +84,7 @@ def generate_prop_defines(node_path, prop):
     elif 'clocks' in prop:
         clocks.extract(node_path, prop, def_label)
-    elif 'pwms' in prop or 'gpios' in prop:
+    elif 'pwms' in prop or '-gpios' in prop or "gpios" == prop:
```
Was that meant to be `prop == "gpios"`? Looks a bit Yoda.

The old code is broken in that `"-gpios" in "foo-gpios-lol"` is true, but it's not a regression.

galak replied:
Fixed up.

ulfalizer reviewed (changes requested):
Fine, except I found a `write_props()` simplification.

galak committed:
scripts/dts/gen_defines: Cleanup write_prop for which props we process

We don't create props for compound types so we dont need to check for
'pwms', 'gpios', 'clocks'.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -154,8 +154,7 @@ def write_props(dev):
         # Skip properties that we handle elsewhere
-        if prop.name in {"reg", "interrupts", "pwms", "clocks", "compatible"} or \
-           prop.name.endswith("gpios"):
+        if prop.name in {"reg", "interrupts", "compatible", "clocks"}:
```
Shouldn't `clocks` also be defined with `type: compound` in `base.yaml`? It also has the format `phandle value phandle value ...` (except `#clock-cells` is usually 0).

galak replied:
Fixed up base binding for clocks to be compound.

galak committed:
dts/bindings: Change clock type to compound

The clock property should be a compound type as that matches uses like
FOO-gpios, pwms, etc.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
